src/utils/promiseAllObject.js

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+import isPlainObject from "./isPlainObject";
+import isFunction from "./isFunction";
+import zipObject from "./zipObject";
+
+/**
+ * Recursively searches through an object or array for any
+ * promises. A new object or array will be composed with
+ * the values of any promises found. This will be used
+ * to resolve the returned promise.
+ *
+ * Note that if a promise returns an object or array, that object
+ * or array will not be recursively searched for promises.
+ *
+ * @param {Object|Array} object
+ * @returns {Promise}
+ */
+const promiseAllObject = object => {
+  const keys = Object.keys(object);
+  return Promise.all(
+    keys.map(key => {
+      const value = object[key];
+      if (
+        Array.isArray(value) ||
+        (isPlainObject(value) && !isFunction(value.then))
+      ) {
+        return promiseAllObject(value);
+      }
+      return value;
```

Comment:
> In the case where an object is passed that was created with a custom constructor, the values of the object are not cloned.

I don't think we want to support custom constructor objects for this function. I don't think it's worth adding a programmatic check, but I do think it's worth adding to the JSDoc. I'll do that in a new commit.

> Additionally, the resolved promise is not cloned. For example, the promise could resolve to an object, and then the object could change before the network call is made.

It's a good point. Let me think about this one. I might make `promiseAllObject` recursively apply to the resolution object as well. I was trying to avoid it because of the extra code it would require, but it might be worth it to also address what you've brought up.

Reply:
I've added a commit intended to improve the jsdoc in this regard. Also, I made `promiseAllObject` recursively search resolved values if they are objects or arrays (which includes cloning them). IMO it actually made the code simpler, so I like the change. Depending on how we decide to log that we're waiting for promises, it might get dirtier again (because we'll need logic to determine whether an object is a promise), but it's good for now.